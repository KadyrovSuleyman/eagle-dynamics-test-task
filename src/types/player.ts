export interface Player {
  id: string,
  status: string,
  name: string,
  avatar: string,
  description?: string,
}
