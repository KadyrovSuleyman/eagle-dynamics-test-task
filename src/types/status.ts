const STATUS = {
  OFFLINE: 'offline',
  ONLINE: 'online',
};

export default Object.freeze(STATUS);
export type StatusT = typeof STATUS;
