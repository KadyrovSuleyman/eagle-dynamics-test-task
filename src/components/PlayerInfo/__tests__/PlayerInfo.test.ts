import { mount, Wrapper, createLocalVue } from '@vue/test-utils';
import STATUS from '@/types/status';
import Vue from 'vue';
import Vuex, { Store } from 'vuex';
import { Player } from '@/types/player';
import PlayerInfoVue from '../PlayerInfo.vue';

jest.mock('../state.ts');
jest.mock('../handlers.ts');

const data = {
  id: 'we0m',
  avatar: 'avatar',
  name: 'name',
  status: STATUS.OFFLINE,
  description: 'description',
};

const newData = {
  id: 'we1m',
  avatar: 'new avatar',
  name: 'new name',
  status: STATUS.ONLINE,
  description: 'new description',
};

let store: Store<{
  selected: boolean,
  player: Player,
}>;

const actions = {
  ban: jest.fn(),
};

beforeEach(() => {
  store = new Vuex.Store({
    state: {
      selected: true,
      player: {
        ...data,
      },
    },
    mutations: {
      changeSelectionState: () => { store.state.selected = !store.state.selected; },
    },
    actions,
  });
});

const localVue = createLocalVue();
localVue.use(Vuex);
let wrapper: Wrapper<Vue>;
beforeEach(() => {
  wrapper = mount(PlayerInfoVue, { localVue, store });
});

// ===============================
it('renders', async () => {
  await wrapper.setProps(data);
  expect(wrapper.element.outerHTML).toMatchSnapshot();
});

it('watchs props change', async () => {
  store.state.player = {
    ...data,
  };
  await wrapper.vm.$nextTick();
  expect(wrapper.element.outerHTML).toMatchSnapshot();

  store.state.player = {
    ...newData,
  };
  await wrapper.vm.$nextTick();
  expect(wrapper.element.outerHTML).toMatchSnapshot();
});

it('detect empty description', async () => {
  store.state.player = {
    ...data,
  };
  await wrapper.vm.$nextTick();
  expect(wrapper.find('.playerInfo-description').text()).toBe(data.description);
  expect(wrapper.find('.playerInfo-emptyDescription').exists()).toBeFalsy();

  store.state.player = {
    ...newData,
    description: undefined,
  };
  await wrapper.vm.$nextTick();
  expect(wrapper.find('.playerInfo-description').exists()).toBeFalsy();
  expect(wrapper.find('.playerInfo-emptyDescription').text())
    .toBe('У этого игрока нет описания');
});

it('watchs outer store', async () => {
  expect(store.state.selected).toBeTruthy();
  expect(wrapper.find('.app-playerInfo').exists()).toBeTruthy();
  expect(wrapper.find('.app-playerInfoFallback').exists()).toBeFalsy();

  store.commit('changeSelectionState');
  await wrapper.vm.$nextTick();
  expect(store.state.selected).toBeFalsy();
  expect(wrapper.find('.app-playerInfo').exists()).toBeFalsy();
  expect(wrapper.find('.app-playerInfoFallback').exists()).toBeTruthy();
});

it('exclude button click triggers his callback', async () => {
  expect(actions.ban).toBeCalledTimes(0);

  await wrapper.find('.playerInfo-kickButton').trigger('click');
  expect(actions.ban).toBeCalledTimes(1);
  expect(actions.ban.mock.calls[0][1]).toBe(data.id);
});
