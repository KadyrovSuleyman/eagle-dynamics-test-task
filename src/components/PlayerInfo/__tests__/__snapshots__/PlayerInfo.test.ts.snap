// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div class="app-playerInfo" id="we0m" avatar="avatar" name="name" status="offline" description="description">
  <div class="playerInfo-avatar"><img src="avatar" alt="name" class="playerInfo-img"></div>
  <div class="playerInfo-name">
    name
  </div>
  <div class="playerInfo-status playerInfo-status__offline">
    offline
  </div>
  <div class="playerInfo-description">
    description
  </div> <button class="playerInfo-kickButton">
    Exclude
  </button>
</div>
`;

exports[`watchs props change 1`] = `
<div class="app-playerInfo">
  <div class="playerInfo-avatar"><img src="avatar" alt="name" class="playerInfo-img"></div>
  <div class="playerInfo-name">
    name
  </div>
  <div class="playerInfo-status playerInfo-status__offline">
    offline
  </div>
  <div class="playerInfo-description">
    description
  </div> <button class="playerInfo-kickButton">
    Exclude
  </button>
</div>
`;

exports[`watchs props change 2`] = `
<div class="app-playerInfo">
  <div class="playerInfo-avatar"><img src="new avatar" alt="new name" class="playerInfo-img"></div>
  <div class="playerInfo-name">
    new name
  </div>
  <div class="playerInfo-status playerInfo-status__online">
    online
  </div>
  <div class="playerInfo-description">
    new description
  </div> <button class="playerInfo-kickButton">
    Exclude
  </button>
</div>
`;
